File: plugins/adaptive-learner-plugin-missions/adaptive_learner_missions/generator.py
"""Daily mission generator - backend (EXP-010 / Phase 56C).

Same algorithm shape as ``frontend/src/lib/missions/generator.ts``:
deterministic per (user_id, date, mix) via a seeded PRNG, adaptive
category eligibility, one pick per difficulty slot, no back-to-back
repeats, only assignable (trackable) checks.

The PRNG differs from the TS one (Python ``random.Random`` vs JS
mulberry32), so exact cross-backend selection is NOT guaranteed -
it does not need to be: a given user uses ONE storage backend
(Dexie OR a shared API DB), and within each backend assignment is
deterministic + idempotent. "Same missions across devices" holds
via the seed (Dexie) or the shared row (API).
"""

from __future__ import annotations

import random

from .catalog import load_templates
from .schema import MissionTemplate

# Parity with frontend/src/lib/missions/checks.ts.
SUPPORTED_CHECK_FUNCTIONS: frozenset[str] = frozenset(
    {
        "lessons_completed_today",
        "lessons_min_2_stars_today",
        "lessons_min_3_stars_today",
        "new_sets_started_today",
        "elements_reviewed_today",
        "elements_mastered_today",
        "perfect_lessons_today",
        "minutes_learned_today",
        "streak_kept_today",
        "current_streak_days",
        "weekend_learning_today",
    }
)

_DIFFICULTY_SEQUENCES: dict[str, list[str]] = {
    "balanced": ["easy", "medium", "hard"],
    "easy": ["easy", "easy", "medium"],
    "challenging": ["medium", "hard", "hard"],
}


def eligible_categories(*, lessons_completed: int, has_errors: bool) -> set[str]:
    if lessons_completed == 0:
        return {"learning", "exploration"}
    cats = {"learning", "exploration", "streak"}
    if has_errors:
        cats.add("review")
        cats.add("mastery")
    return cats


def assign_daily_missions(
    user_id: str,
    date_iso: str,
    *,
    lessons_completed: int,
    has_errors: bool,
    is_weekend: bool,
    count: int = 3,
    difficulty_mix: str = "balanced",
    exclude_ids: tuple[str, ...] = (),
) -> list[MissionTemplate]:
    count = max(1, min(3, count))
    mix = difficulty_mix if difficulty_mix in _DIFFICULTY_SEQUENCES else "balanced"
    exclude = set(exclude_ids)
    rng = random.Random(f"{user_id}:{date_iso}:{mix}")

    cats = eligible_categories(lessons_completed=lessons_completed, has_errors=has_errors)
    eligible = [
        t
        for t in load_templates()
        if t.check_function in SUPPORTED_CHECK_FUNCTIONS
        and t.category.value in cats
        and t.id not in exclude
        and not (t.id == "weekend-learner" and not is_weekend)
    ]

    sequence = _DIFFICULTY_SEQUENCES[mix][:count]
    picked: list[MissionTemplate] = []
    picked_ids: set[str] = set()

    def pick_from(pool: list[MissionTemplate]) -> MissionTemplate | None:
        for t in pool:
            if t.id not in picked_ids:
                return t
        return None

    for difficulty in sequence:
        order = list(eligible)
        rng.shuffle(order)
        chosen = pick_from([t for t in order if t.difficulty.value == difficulty]) or pick_from(
            order
        )
        if chosen is not None:
            picked.append(chosen)
            picked_ids.add(chosen.id)

    return picked

File: frontend/src/lib/missions/generator.ts
/**
 * Daily mission generator (EXP-010 / Phase 56C / P-157, P-162).
 *
 * Deterministic: the same user sees the same missions on the same
 * day across devices, because selection is driven by a PRNG seeded
 * from ``user_id + date``. Difficulty + category eligibility adapt
 * to the learner's history (new / active / veteran), and the same
 * mission is never assigned two days running.
 *
 * Pure - no storage access. The caller supplies the profile +
 * yesterday's template ids; the missions namespace persists the
 * result.
 */

import {getTemplates} from "./catalog";
import {isSupportedCheck} from "./checks";
import type {
    DifficultyMix,
    MissionDifficulty,
    MissionProfile,
    MissionTemplate,
} from "./types";

/** FNV-1a 32-bit hash → seed. */
function hashSeed(input: string): number {
    let h = 0x811c9dc5;
    for (let i = 0; i < input.length; i++) {
        h ^= input.charCodeAt(i);
        h = Math.imul(h, 0x01000193);
    }
    return h >>> 0;
}

/** mulberry32 PRNG: deterministic, well-distributed, tiny. */
function mulberry32(seed: number): () => number {
    let a = seed >>> 0;
    return () => {
        a |= 0;
        a = (a + 0x6d2b79f5) | 0;
        let t = Math.imul(a ^ (a >>> 15), 1 | a);
        t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
        return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
}

const DIFFICULTY_SEQUENCES: Record<DifficultyMix, MissionDifficulty[]> = {
    balanced: ["easy", "medium", "hard"],
    easy: ["easy", "easy", "medium"],
    challenging: ["medium", "hard", "hard"],
};

/** Which categories a learner is eligible for, by history. */
export function eligibleCategories(profile: MissionProfile): Set<string> {
    // A brand-new user (no completed lessons) gets only the
    // gentle, always-doable categories.
    if (profile.lessonsCompleted === 0) {
        return new Set(["learning", "exploration"]);
    }
    const cats = new Set(["learning", "exploration", "streak"]);
    // Review + mastery only make sense once the learner has error
    // history to act on.
    if (profile.hasErrors) {
        cats.add("review");
        cats.add("mastery");
    }
    return cats;
}

/** Deterministic Fisher-Yates using the supplied rng. */
function shuffled<T>(items: readonly T[], rng: () => number): T[] {
    const out = items.slice();
    for (let i = out.length - 1; i > 0; i--) {
        const j = Math.floor(rng() * (i + 1));
        [out[i], out[j]] = [out[j], out[i]];
    }
    return out;
}

export interface AssignOptions {
    count?: number; // 1..3, default 3
    difficultyMix?: DifficultyMix; // default "balanced"
    /** Template ids assigned yesterday (never repeat back-to-back). */
    excludeIds?: readonly string[];
}

/**
 * Pick the day's missions for a user. Deterministic per
 * (userId, dateISO, profile, options).
 */
export function assignDailyMissions(
    userId: string,
    dateISO: string,
    profile: MissionProfile,
    options: AssignOptions = {},
): MissionTemplate[] {
    const count = Math.max(1, Math.min(3, options.count ?? 3));
    const mix = options.difficultyMix ?? "balanced";
    const exclude = new Set(options.excludeIds ?? []);
    const rng = mulberry32(hashSeed(`${userId}:${dateISO}:${mix}`));

    const cats = eligibleCategories(profile);
    const eligible = getTemplates().filter((t) => {
        // Only assign missions whose progress we can actually
        // track against existing data.
        if (!isSupportedCheck(t.check_function)) return false;
        if (!cats.has(t.category)) return false;
        if (exclude.has(t.id)) return false;
        // The weekend-learner mission is only offered on weekends.
        if (t.id === "weekend-learner" && !profile.isWeekend) return false;
        return true;
    });

    const sequence = DIFFICULTY_SEQUENCES[mix].slice(0, count);
    const picked: MissionTemplate[] = [];
    const pickedIds = new Set<string>();

    const pickFrom = (pool: MissionTemplate[]): MissionTemplate | null => {
        for (const t of pool) {
            if (!pickedIds.has(t.id)) return t;
        }
        return null;
    };

    for (const difficulty of sequence) {
        const order = shuffled(eligible, rng);
        // Prefer the requested difficulty; fall back to any
        // eligible template when that bucket is exhausted (small
        // catalogs / restricted new-user eligibility).
        const chosen =
            pickFrom(order.filter((t) => t.difficulty === difficulty)) ??
            pickFrom(order);
        if (chosen) {
            picked.push(chosen);
            pickedIds.add(chosen.id);
        }
    }

    return picked;
}
